### PR TITLE
*: compatible upgrade with previous versions

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/46275

Problem Summary:
* `version < v7.1.0`, we don't support upgrading TiDB without manually Canceling DDL
* `v7.4.0 > version >= v7.1.0`, we support upgrading TiDB without manually Canceling DDL, but we don't support `/upgrade/start` and `upgrade/finish` APIs.
* `version >= v7.4.0`, we support `/upgrade/start` and `upgrade/finish` APIs.

### What is changed and how it works?
* For compatibility:
    * When upgrading from the version in the `v7.4.0 > version >= v7.1.0` range, we do not need to execute `/upgrade/start` before upgrading.
    * When upgrading from the `version >= v7.4.0` version, we need to execute `/upgrade/start` before upgrading.
   
* ddl/mock.go : We move `clusterState` out of `MockStateSyncer`. Because we want to make it unaffected by ddl close.
* Update `modifyBootstrapVersionForTest`
    * Remove `modifyBootstrapVersionForTest` code, because IT removes this logic.
    * Now we want to test `SupportUpgradeHTTPOpVer` upgrade `SupportUpgradeHTTPOpVer++`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
When upgrading from a version greater than or equal to v7.4.0, an explicit `/upgrade/start` HTTP request is required to start the cluster upgrade
```

### CODE REVIEW

File: ddl/mock.go [hunk 1]
```diff
@@ -148,43 +148,48 @@ func NewMockStateSyncer() syncer.StateSyncer {
 	return &MockStateSyncer{}
 }
 
+// clusterState mocks cluster state.
+// We move it from MockStateSyncer to here. Because we want to make it unaffected by ddl close.
+var clusterState *atomicutil.Pointer[syncer.StateInfo]
```
Comment (Benjamin2037) [inline ddl/mock.go:153]: will this clusterState changed in concurrency way？

File: server/handler/upgrade_handler.go [hunk 2]
```diff
@@ -46,9 +46,9 @@ func (h ClusterUpgradeHandler) ServeHTTP(w http.ResponseWriter, req *http.Reques
 	op := req.FormValue("op")
 	switch op {
 	case "start":
-		hasDone, err = h.startUpgrade()
+		hasDone, err = h.StartUpgrade()
```
Comment (Benjamin2037) [inline server/handler/upgrade_handler.go:49]: If there is err returned，how to handle it？

File: ddl/mock.go (hunk 1 above)
Comment (zimulala) [reply ddl/mock.go:153]: It's possible(call `UpdateGlobalState` in parallel), so it's used `atomicutil`.

File: server/handler/upgrade_handler.go (hunk 2 above)
Comment (zimulala) [reply server/handler/upgrade_handler.go:49]: Return error directly.

Comment (zimulala) [reply server/handler/upgrade_handler.go:49]: Try again, just like in bootstrap

File: server/handler/upgrade_handler.go
Comment (Benjamin2037) [inline server/handler/upgrade_handler.go:1]: line 38 ServeHTTP， is ServerHTTP？

Comment (zimulala) [reply server/handler/upgrade_handler.go:1]: Yes.

File: session/bootstrap.go [hunk 3]
```diff
@@ -1330,6 +1332,24 @@ func IsUpgradingClusterState(s Session) (bool, error) {
...
 
+func checkOrSyncUpgrade(s Session, ver int64) {
+	if ver < SupportUpgradeHTTPOpVer {
+		terror.MustNil(SyncUpgradeState(s))
+		return
+	}
+
+	isUpgrading, err := IsUpgradingClusterState(s)
+	if err != nil {
+		logutil.BgLogger().Fatal("get global state failed", zap.String("category", "upgrading"), zap.Error(err))
```
Comment (Benjamin2037) [inline session/bootstrap.go:1343]: Should we return to up-caller is better than just directly Fatal?

File: server/handler/upgrade_handler.go
Comment (Benjamin2037) [reply server/handler/upgrade_handler.go:1]: Then please update it

Comment (zimulala) [reply server/handler/upgrade_handler.go:1]: I'm sorry, that's just `ServeHTTP`. I see all the other interfaces use it.

**Comment by zimulala** [comment]
PTAL @Benjamin2037 @tangenta

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: session/bootstrap.go [hunk 4]
```diff
@@ -1330,6 +1334,36 @@ func IsUpgradingClusterState(s Session) (bool, error) {
 	return stateInfo.State == syncer.StateUpgrading, nil
 }
 
+func checkOrSyncUpgrade(s Session, ver int64) {
+	if ver < SupportUpgradeHTTPOpVer {
+		terror.MustNil(SyncUpgradeState(s))
+		return
```
Comment (tangenta) [inline session/bootstrap.go:1340]: No need to set `SyncNormalRunning`?

**Review by tangenta — APPROVED** [review]

File: session/bootstrap.go (hunk 4 above)
Comment (zimulala) [reply session/bootstrap.go:1340]: Notify the cluster that the upgrade is complete by occurring `upgrade/finish`, and this request will send to the new TiDB(>= v7.4)

**Comment by easonn7** [comment]
/approve

There is no sysvar change.

**Pull request merged**